### PR TITLE
pk: make private and public key private structures

### PR DESCRIPTION
**Pull request by hannesm**
validate before constructing public and private keys. use result instead of exceptions in of_exp, of_prime

remove well_formed, which is now implicit.

### CODE REVIEW

File: pk/mirage_crypto_pk.mli [hunk 1]
```diff
@@ -149,16 +122,17 @@ module Rsa : sig
 
   (** {1 Key generation} *)
 
-  val generate : ?g:Mirage_crypto_rng.g -> ?e:Z.t -> bits -> priv
-  (** [generate g e bits] is a new {{!priv}private key}. The new key is
+  val generate : ?g:Mirage_crypto_rng.g -> ?e:Z.t -> bits:bits -> unit -> priv
+  (** [generate ~g ~e ~bits ()] is a new {{!priv}private key}. The new key is
       guaranteed to be {{!well_formed}well formed}.
```
Comment (cfcs) [inline pk/mirage_crypto_pk.mli]:
```suggestion
      guaranteed to be well-formed (see {!priv} and {!pub}).
```

File: pk/rsa.ml [hunk 2]
```diff
@@ -35,48 +26,147 @@ let digest_or ~hash =
...
+
+type pub = { e : Z_sexp.t ; n : Z_sexp.t } [@@deriving sexp]
+
+(* due to PKCS1 *)
+let minimum_octets = 12
+let minimum_bits = 8 * minimum_octets - 7
+
+let guard p err = if p then Ok () else Error err
+
+open Rresult.R.Infix
```
Comment (dinosaure) [inline pk/rsa.ml]: `open Rresult` defines infix operators. The idea is to do
```ocaml
open Rresult

(* free to use ( >>= ) and ( >>| ) and free to use [R] module. *)
let _ =
   foo x >>= bar x >>= fun () -> R.ok true
```

`Infix` module comes to mix `rresult` operator with some others like `lwt`.

File: pk/rsa.ml [hunk 3]
```diff
@@ -35,48 +26,147 @@ let digest_or ~hash =
...
+  | Error (`Msg m) -> failwith "bad public key: %s" m
+
+type priv = {
+  e : Z_sexp.t ; d : Z_sexp.t ; n  : Z_sexp.t ;
+  p : Z_sexp.t ; q : Z_sexp.t ; dp : Z_sexp.t ; dq : Z_sexp.t ; q' : Z_sexp.t
+} [@@deriving sexp]
+
+let valid_prime name p =
+  guard Z.(p > zero && is_odd p && Z_extra.pseudoprime p)
+    (`Msg ("invalid prime " ^ name))
```
Comment (dinosaure) [inline pk/rsa.ml]: `R.msgf` is better I think.

File: pk/rsa.ml [hunk 4]
```diff
@@ -35,48 +26,147 @@ let digest_or ~hash =
...
+  match priv ~e:p.e ~d:p.d ~n:p.n ~p:p.p ~q:p.q ~dp:p.dp ~dq:p.dq ~q':p.q' with
+  | Error (`Msg m) -> failwith "invalid private key %s" m
+  | Ok p -> p
+
 let priv_of_primes ~e ~p ~q =
+  valid_prime "p" p >>= fun () ->
+  valid_prime "q" q >>= fun () ->
+  guard (p <> q) (`Msg "p and q are the same prime") >>= fun () ->
   let n  = Z.(p * q)
   and d  = Z.(invert e (pred p * pred q)) in
```
Comment (cfcs) [inline pk/rsa.ml]: In `priv` we have
`guard Z.(one < d && d < n) (`Msg "invalid private exponent") >>= fun () ->`
Is that also guaranteed here?

File: pk/rsa.ml (hunk 2 above)
Comment (cfcs) [reply pk/rsa.ml]: We're not using Lwt in this module though, so is that a problem?

Comment (dinosaure) [reply pk/rsa.ml]: When you `open Rresult`, you are able to use `R....`. Currently, if you want to use something like `R.msgf`, you must fully qualified `Rresult.R.msgf`. The advise comes with my other comment [to prefer `R.msgf` instead to use a concatenation](https://github.com/mirage/mirage-crypto/pull/28#discussion_r388303199).

File: pk/rsa.ml (hunk 4 above)
Comment (hannesm) [reply pk/rsa.ml]: yes, indeed (from https://gmplib.org/manual/Number-Theoretic-Functions.html#Number-Theoretic-Functions):
Compute the inverse of op1 modulo op2 and put the result in rop. If the inverse exists, the return value is non-zero and rop will satisfy 0 <= rop < abs(op2) (with rop = 0 possible only when abs(op2) = 1, i.e., in the somewhat degenerate zero ring). If an inverse doesn’t exist the return value is zero and rop is undefined. The behaviour of this function is undefined when op2 is zero.

--> d will always be positive and smaller the provided modulus (here (p - 1) * (q - 1) -- which is < n).

**Comment by hannesm** [comment]
I have now also adapted DSA with private structures, and validation of public and private keys. In addition, I adapted how OpenSSL blinds DSA signing, and removed a non-termination from `Dsa.sign` (now leading to an Invalid_argument - happy to hear better solutions here, see discussion in https://tools.ietf.org/html/rfc6979#section-3.4 - I believe exceptional behaviour is better than non-termination). This PR is now good to be reviewed and merged. Afterwards we can finally use `powm_sec` (which required the modulus to be odd and the exponent to be positive) without potentially triggering exceptions (happy to do this as part of this PR or in a subsequent one).

**Pull request merged**